h5netcdf/core.py

```diff
@@ -646,16 +588,35 @@ def _create_child_variable(
                 self._root.invalid_netcdf,
             )
 
+        # maybe create new dimensions depending on data
+        if data is not None:
+            for d, s in zip(dimensions, data.shape):
+                # create new dimensions only ever if
+                #  - they are not known via parent-groups
+                #  - they are given in dimensions
+                #  - it's not a coordinate variable, they will get special handling later
+                if d not in self._all_dimensions and d in dimensions and d is not name:
+                    # calls _create_dimension
+                    self.dimensions[d] = s
```

Comment:
This will still create invalid state if an exception is raised later, right? Maybe it would be worth wrapping this in `try ... except` or so?

Reply:
True, if a dimension is created here and something breaks afterwards, this would be the consequence.

I've used your example as test. At least this case is fixed. But you are right, we should take care of likely breakages.

Do you can think of something  which breaks with this implementation? Refactoring this function is still on my personal todo list.

Reply:
For example users wanting to use `chunks="h5netcdf"`, but not knowing whether it's available:

```python
try:
   f.create_variable(..., chunks="h5netcdf")
except:
   # fall back to h5py chunking
   f.create_variable(..., chunks=True)

# oops, file is broken
```

Or more general, pretty much every internal error in `h5py`'s `create_dataset`.

Seem like a somewhat simple fix would be this:

```python
orig_dimensions = self.dimensions.copy()

try:
    ...
except:
    # roll back changes
    self.dimensions = orig_dimensions
```

Reply:
👍  if at all possible, it is much more user friendly to make "atomic" functions.

Reply:
@dionhaefner @shoyer I'd suggest to tackle that problem in a follow up PR. This one here is already quite huge. 

Issue #138 exist and we can coordinate there.